runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_i8mm.c

```diff
@@ -107,3 +107,159 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0(
+// This has the effect of multiplying each int4 by 2^4 = 16. To compensate,
+// we divide the accumulator values by 16 before storing to memory.
+// This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+// kernels in https://github.com/google/XNNPACK.
+
+IREE_UK_ATTRIBUTE_ALWAYS_INLINE inline void
+iree_uk_mmt4d_tile_s8s4s32_1x8x16_arm_64_i8mm(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params) {
+  IREE_UK_ASSERT(!(params->K0 % 16));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int8x8_t vzero = vmov_n_s8(0);
+
+  int32x4_t acc[4];
+  IREE_UK_UNROLL for (int i = 0; i < 4; i++) {
+    // We start with zero accumulators and add the value of *out_ptr later.
+    // This is required for the int4 left shift described above.
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t rhs[2][4];
+    IREE_UK_UNROLL for (int i = 0; i < 4; i++) {
+      int8x16_t r = vld1q_s8(rhs_ptr + 16 * i);
+      rhs[0][i] = vshlq_n_s8(r, 4);
+      rhs[1][i] = vandq_s8(r, vmask);
+    }
+    rhs_ptr += 64;
+
+    int8x16_t lhs[2];
+    int8x8x2_t lhs_uzp = vld2_s8(lhs_ptr);
+    lhs_ptr += 16;
+    lhs[0] = vcombine_s8(lhs_uzp.val[0], vzero);
+    lhs[1] = vcombine_s8(lhs_uzp.val[1], vzero);
+
+    IREE_UK_UNROLL for (int i = 0; i < 4; i++) {
+      acc[i] = vmmlaq_s32(acc[i], lhs[0], rhs[0][i]);
+      acc[i] = vmmlaq_s32(acc[i], lhs[1], rhs[1][i]);
+    }
+  }
+
+  IREE_UK_UNROLL for (int j = 0; j < 2; j++) {
+    acc[2 * j + 0] = vshrq_n_s32(acc[2 * j + 0], 4);
+    acc[2 * j + 1] = vshrq_n_s32(acc[2 * j + 1], 4);
+
+    int32x4_t acc_1x4_0 =
+        iree_uk_neon_uzp1_s32_as_s64(acc[2 * j + 0], acc[2 * j + 1]);
+    if (params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE) {
+      int32x4_t existing_acc = vld1q_s32(out_ptr + 4 * j);
+      acc_1x4_0 = vaddq_s32(acc_1x4_0, existing_acc);
+    }
+    vst1q_s32(out_ptr + 4 * j, acc_1x4_0);
+  }
+}
+
+IREE_UK_ATTRIBUTE_ALWAYS_INLINE static inline void
+iree_uk_mmt4d_tile_s8s4s32_2x8x16_to_8x8x16_arm_64_i8mm(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 2 && M0 <= 8 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 16));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int mtiles = M0 / 2;
+
+  int32x4_t acc[4][4];
+  for (int i = 0; i < mtiles; i++) {
```

Comment:
IREE_UK_UNROLL. As mtiles = M0/2 and M0 is a constant in this always-inlined function, so is mtiles.